### PR TITLE
Add API-based support for [GitLab] badges, add new GitLab Tag badge

### LINKED ISSUE #541: Badge request: Gitlab CI and Gitlab
It would be awesome if Shields supported Gitlab and Gitlab CI. They have a Decent API and it should be 100% feasible.

### PR DESCRIPTION
**Pull request by calebcartwright**
Closes #541, refs/implements the tag half of #4867. I've opted to use the close keyword on #541 as I feel that with the authenticated supported added here, and accompanying rate limit increase, we can now move forward with other, specific badge requests as needed following our standard badge intake flow.

Have already setup the account and token, will plug them into the prod env before deploying and will also share the relevant account info with the rest of the ops team.

### CODE REVIEW

File: services/gitlab/gitlab-base.js
Comment (calebcartwright) [inline services/gitlab/gitlab-base.js:6]: We'll constrain the allowed origins for Shields.io to `gitlab.com`, but need this to be flexible for other self-hosters which is why the key is used here instead of hardcoding the `authorizedOrigins` value to `gitlab.com`

File: doc/server-secrets.md
Comment (chris48s) [inline doc/server-secrets.md:153]: Minor copy & paste errors in the docs here (drone)

File: services/gitlab/gitlab-tag.service.js
Comment (chris48s) [inline services/gitlab/gitlab-tag.service.js:14]: The `gitlab_url` param isn't showing up in the builder/builder modal in the front-end. I think we need a second example for the case where `gitlab_url` is set to enable that.

File: services/gitlab/gitlab-tag.service.js
Comment (chris48s) [inline services/gitlab/gitlab-tag.service.js:59]: This is not the latest tag by date. From the docs https://docs.gitlab.com/ee/api/tags.html

> Get a list of repository tags from a project, sorted by name in reverse alphabetical order

Not checked, but I think this is basically the same situation we've got with the GitHub API so we can maybe copy the sort by date logic from there?

File: services/gitlab/gitlab-tag.spec.js
Comment (chris48s) [inline services/gitlab/gitlab-tag.spec.js:24]: My reading of the code here is that we'll send our token to gitlab.com (or whatever is in `GITLAB_ORIGINS`), but you can still specify a custom `gitlab_url` (that is not in `GITLAB_ORIGINS`) and it will work (e.g: https://shields-staging-pr-6988.herokuapp.com/gitlab/tag/GNOME/pango?gitlab_url=https%3A%2F%2Fgitlab.gnome.org does render a badge) but it just won't send the token for those requests. Two follow-ups for that:
1. Can we add a second test case here that explicitly verifies our token is not sent when we specify a custom `gitlab_url` (and the url is not in `GITLAB_ORIGINS`)
2. Do we think the anonymous rate limit becomes a problem for alternate gitlab installs? Presumably self-hosters can configure this differently to the centralised service.. Do you think it is necessary to allow users to also pass a `&token=` param? Maybe we can just ignore this to start with it and deal with if it becomes a problem. I'm just trying to think ahead a bit..

File: services/gitlab/gitlab-tag.tester.js
Comment (chris48s) [inline services/gitlab/gitlab-tag.tester.js:15]: Can we add another test for the custom `gitlab_url` case here too please.

**Review by chris48s — COMMENTED** [review]
Nice work on tackling this. One question here. The rest are inline on the diff.

> Have already setup the account and token, will plug them into the prod env before deploying

Do we also need to set a token for staging and/or CI or will the anonymous rate limit be sufficient for those environments?

**Comment by calebcartwright** [comment]
>  Do we also need to set a token for staging and/or CI or will the anonymous rate limit be sufficient for those environments?

I was thinking anonymous will be fine. With a 2k per minute limit I don't anticipate any issues in staging nor CI

File: services/gitlab/gitlab-tag.service.js
Comment (calebcartwright) [reply services/gitlab/gitlab-tag.service.js:59]: I haven't actually really looked at any part of the tag implementation, just grabbed the commits from #6389 as had been discussed in #541. Will need take a closer look

**Comment by chris48s** [comment]
> 2k per minute

Right, yeah that is pretty generous and I think that should cover the custom `gitlab_url` case for most users :+1:

File: services/gitlab/gitlab-tag.spec.js
Comment (calebcartwright) [reply services/gitlab/gitlab-tag.spec.js:24]: > Can we add a second test case here that explicitly verifies our token is not sent when we specify a custom gitlab_url (and the url is not in GITLAB_ORIGINS)

I pulled in this test case because we have one such case for all other authenticated services. I guess I'm not strongly opposed but we don't have it for any of the similar services which I imagine is because the negative case is sufficiently covered in the auth helper tests, such as:

https://github.com/badges/shields/blob/f4bddb9964247a8d83803b3a59c13454e33c9cc6/core/base-service/auth-helper.spec.js#L344-L363

> Do we think the anonymous rate limit becomes a problem for alternate gitlab installs? Presumably self-hosters can configure this differently to the centralised service.. Do you think it is necessary to allow users to also pass a &token= param?

My reading of the GitLab docs on throttling is that it is IP based, so I don't think allowing users to provide a token nor even a hypothetical GitHub-like token pooling solution would be of any help for us on the rate limiting front.

GitLab does support scoped access tokens, including the ability to scope in a strictly read-only way. As such I think allowing users to provide their own token could help with cases of badges for private projects via Shields.io without having to self-host the badge server. However, I suspect that will pose a technical problem with our current auth injection mechanisms, since we'd need to pivot to a "don't worry about the authorized origins if it's a user provided token, but still enforce authorized origins with the configured token if one wasn't provided". Presumably a problem that can be overcome, but probably best done in a separate dedicated thread.

I don't think we'll have the ability to readily support authenticated requests for alternate instances without another type of non-trivial change to our auth story, as we'd need the ability to define multiple mappings of an auth mechanism to a set of authorized origins, whereas today it's strictly a 1 secret/token/creds to many origins.

File: services/gitlab/gitlab-tag.service.js
Comment (calebcartwright) [reply services/gitlab/gitlab-tag.service.js:59]: I feel like their documentation is somewhat contradictory :thinking: The textual description of the API shared above is rather unequivocal about the sort being alphabetical by name, but then the param documentation just a couple lines below says:

> order_by | string | no | Return tags ordered by name or updated fields. Default is updated

I'm going to need to setup a test repo to try to vet it out because the handful of real repositories I was trying to spot check with were inconclusive due to both means producing the same ordering

Comment (calebcartwright) [reply services/gitlab/gitlab-tag.service.js:59]: All signs point to the default sort indeed being by date, despite what the API description says.

https://gitlab.com/api/v4/projects/shields-ops-group%2Ftag-test/repository/tags

<details><summary>API response</summary>

```json
[
  {
    "name": "v2.0.0",
    "message": "",
    "target": "ae4abdc07357a1d26d3c692a9ad3118b6b31af2a",
    "commit": {
      "id": "ae4abdc07357a1d26d3c692a9ad3118b6b31af2a",
      "short_id": "ae4abdc0",
      "created_at": "2021-09-11T11:20:31.000-05:00",
      "parent_ids": [
        "396bf1debee4d196f85ae7f3d00ce2a4a71346a3"
      ],
      "title": "foo",
      "message": "foo\n",
      "author_name": "Caleb Cartwright",
      "author_email": "caleb.cartwright@outlook.com",
      "authored_date": "2021-09-11T11:20:31.000-05:00",
      "committer_name": "Caleb Cartwright",
      "committer_email": "caleb.cartwright@outlook.com",
      "committed_date": "2021-09-11T11:20:31.000-05:00",
      "trailers": {},
      "web_url": "https://gitlab.com/shields-ops-group/tag-test/-/commit/ae4abdc07357a1d26d3c692a9ad3118b6b31af2a"
    },
    "release": null,
    "protected": false
  },
  {
    "name": "v4.0.0",
    "message": "",
    "target": "396bf1debee4d196f85ae7f3d00ce2a4a71346a3",
    "commit": {
      "id": "396bf1debee4d196f85ae7f3d00ce2a4a71346a3",
      "short_id": "396bf1de",
      "created_at": "2021-09-11T11:16:33.000-05:00",
      "parent_ids": [
        "bf047705d78e939fcd7ff5256014c7fcee954a6f"
      ],
      "title": "another update",
      "message": "another update\n",
      "author_name": "Caleb Cartwright",
      "author_email": "caleb.cartwright@outlook.com",
      "authored_date": "2021-09-11T11:16:33.000-05:00",
      "committer_name": "Caleb Cartwright",
      "committer_email": "caleb.cartwright@outlook.com",
      "committed_date": "2021-09-11T11:16:33.000-05:00",
      "trailers": {},
      "web_url": "https://gitlab.com/shields-ops-group/tag-test/-/commit/396bf1debee4d196f85ae7f3d00ce2a4a71346a3"
    },
    "release": null,
    "protected": false
  },
  {
    "name": "v3.0.0",
    "message": "",
    "target": "bf047705d78e939fcd7ff5256014c7fcee954a6f",
    "commit": {
      "id": "bf047705d78e939fcd7ff5256014c7fcee954a6f",
      "short_id": "bf047705",
      "created_at": "2021-09-11T11:11:50.000-05:00",
      "parent_ids": [
        "490f9e3d62ac21d5f64d53b6b8d607d0fd9792d1"
      ],
      "title": "more updates",
      "message": "more updates\n",
      "author_name": "Caleb Cartwright",
      "author_email": "caleb.cartwright@outlook.com",
      "authored_date": "2021-09-11T11:11:50.000-05:00",
      "committer_name": "Caleb Cartwright",
      "committer_email": "caleb.cartwright@outlook.com",
      "committed_date": "2021-09-11T11:11:50.000-05:00",
      "trailers": {},
      "web_url": "https://gitlab.com/shields-ops-group/tag-test/-/commit/bf047705d78e939fcd7ff5256014c7fcee954a6f"
    },
    "release": null,
    "protected": false
  },
  {
    "name": "v1.0.0",
    "message": "",
    "target": "490f9e3d62ac21d5f64d53b6b8d607d0fd9792d1",
    "commit": {
      "id": "490f9e3d62ac21d5f64d53b6b8d607d0fd9792d1",
      "short_id": "490f9e3d",
      "created_at": "2021-09-11T11:10:09.000-05:00",
      "parent_ids": [
        "922ab7ddafd8b7892cfa881f6d67cdde48f59ed1"
      ],
      "title": "update",
      "message": "update\n",
      "author_name": "Caleb Cartwright",
      "author_email": "caleb.cartwright@outlook.com",
      "authored_date": "2021-09-11T11:10:09.000-05:00",
      "committer_name": "Caleb Cartwright",
      "committer_email": "caleb.cartwright@outlook.com",
      "committed_date": "2021-09-11T11:10:09.000-05:00",
      "trailers": {},
      "web_url": "https://gitlab.com/shields-ops-group/tag-test/-/commit/490f9e3d62ac21d5f64d53b6b8d607d0fd9792d1"
    },
    "release": null,
    "protected": false
  }
]
```

</details>

File: doc/server-secrets.md
Comment (calebcartwright) [reply doc/server-secrets.md]: Fixed in ee81adcc8bce2f271d444ed6f144c44f9dff6755

File: services/gitlab/gitlab-tag.service.js
Comment (calebcartwright) [reply services/gitlab/gitlab-tag.service.js]: Fixed in 8d5cfe16691522fa703bafa16e90d0a5c3cb797f

File: services/gitlab/gitlab-tag.service.js
Comment (calebcartwright) [reply services/gitlab/gitlab-tag.service.js]: Feel like this has been sufficiently addressed in b96f597e92c1de7a9857cee5f8720c4533904382 so going to resolve, but let me know if you think there's still any outstanding items on this topic

File: services/gitlab/gitlab-tag.tester.js
Comment (calebcartwright) [reply services/gitlab/gitlab-tag.tester.js:27]: Addressed in b96f597e92c1de7a9857cee5f8720c4533904382

File: services/gitlab/gitlab-tag.service.js
Comment (calebcartwright) [inline services/gitlab/gitlab-tag.service.js:83]: This is mostly copy/pasta from the corresponding GitHub service for consistency, but probably worth a moment of discussion given the change (the introduction of the `v` prefix). 

There's been recurring discussion on this in general, and we certainly want to keep our standard posture (refs #3200, #1181, and #6135). However, we have previously expressed an openness to consider no longer inserting the `v` prefix on the GitHub Tag badges (e.g. https://github.com/badges/shields/issues/3200#issuecomment-535578388).

Accordingly I actually think the message should be adjusted here to just display `version`, sans the `addv` call, but wanted to see what others thought first

File: services/gitlab/gitlab-tag.spec.js
Comment (chris48s) [reply services/gitlab/gitlab-tag.spec.js:24]: > we don't have it for any of the similar services which I imagine is because the negative case is sufficiently covered in the auth helper tests

OK yeah I think its covered by the tests for the base class.

> I don't think we'll have the ability to readily support authenticated requests for alternate instances without another type of non-trivial change to our auth story, as we'd need the ability to define multiple mappings of an auth mechanism to a set of authorized origins, whereas today it's strictly a 1 secret/token/creds to many origins.

Useful to have thought it through a bit, but I think we can just ignore it for now and cross this bridge if it ever becomes an issue. I think the anon rate limit should cover most custom server cases anyway.

File: services/gitlab/gitlab-tag.service.js
Comment (chris48s) [reply services/gitlab/gitlab-tag.service.js]: Odd that the docs are wrong, but I agree your assessment of the real world behaviour :+1:

File: services/gitlab/gitlab-tag.service.js
Comment (chris48s) [inline services/gitlab/gitlab-tag.service.js:37]: This example doesn't match the title. This should be `queryParams: { sort: 'date' }` or just nothing

File: services/gitlab/gitlab-tag.service.js
Comment (chris48s) [inline services/gitlab/gitlab-tag.service.js:72]: The objects we're passing to `render()` in the examples here don't include the right keys so all the examples are showing as ` tag | undefined`

File: services/gitlab/gitlab-tag.service.js
Comment (chris48s) [reply services/gitlab/gitlab-tag.service.js:83]: I do see where you're coming from and this is a good point to raise.

Our starting point for this is that (pretty much?) all of our badges add the v prefix by default. If we were to decide to allow users to disable it, the pattern is going to be that everything appends v by default but then you can pass `?addv=false` or whatever to disable it.

Given that, I think it probably makes most sense to set GitLab up the same way as everything else to start with (i.e: with `addv()`). If we don't we will either end up saying "all the version badges append v by default and you can pass `?addv=false`, except GitLab which is the other way round" or we have to make a behaviour change for GitLab in future to make it consistent.

Comment (calebcartwright) [reply services/gitlab/gitlab-tag.service.js:83]: >  If we were to decide to allow users to disable it, the pattern is going to be that everything appends v by default but then you can pass ?addv=false or whatever to disable it.

to clarify, i'm not proposing we expose this level of control to the users via a query param or other means. it's strictly a question of what Shields does internally, either always injecting a leading `v` or not. I don't think this is something we can change our minds on down the road, or at least it's not something I think we should bank on having the liberty to do, so I feel like we should be confident and comfortable with whatever option we select.

in hindsight the tag badge is somewhat of a challenge to incorporate as the example part of what is largely centered around the general auth story for gitlab, given all the associated nuances

File: services/gitlab/gitlab-tag.service.js
Comment (calebcartwright) [inline services/gitlab/gitlab-tag.service.js:25]: on the recurring theme of consistency, both with github and at large, i've updated the route to match the github tag badge route. however, i realize i'm not completely convinced this aligns with the `service/noun/parameters` badge route pattern enumerated in our docs.

we've also got other github routes that have the `v` and `noun` order reversed, e.g. `github/v/tag/...` vs `github/package-json/v/...`

i think this is probably the best bet to provide the highest degree of consistency, but also feel like maybe the docs could use more clarity if this is a common exception to the main rule?

File: services/gitlab/gitlab-tag.service.js
Comment (calebcartwright) [reply services/gitlab/gitlab-tag.service.js:72]: ugh. good catch, sorry for the noise on this one. too many moving parts and keep missing little things

File: services/gitlab/gitlab-tag.service.js
Comment (chris48s) [reply services/gitlab/gitlab-tag.service.js:25]: Agree with your decision here :+1:

File: services/gitlab/gitlab-tag.service.js
Comment (chris48s) [reply services/gitlab/gitlab-tag.service.js:83]: Yes I think once we pick an approach here we're basically stuck with it. There isn't a single unambiguously correct choice. I think I'm still in favour of `addv()` just because it is what we do on all the other version badges (which this is https://github.com/badges/shields/pull/6988/files#diff-36f70802cce1e0dcc5c595f7913c44e5bce18d58c8357d8567b582399c7fcdb1R22 ) and it is what we'd be doing if we were using `renderVersionBadge()` but if you (or other @badges/shields-maintainers ) reckon there's a strong case for it here I'm not wedded to that opinion. Tags are slightly different and as I say, neither option is unambiguously correct.

**Review by chris48s — DISMISSED** [review]
Pressing the big green tick on this as if we stick with `addv()` it is good to go, but happy to re-:+1: it if we change our minds on https://github.com/badges/shields/pull/6988#discussion_r706641452

File: services/gitlab/gitlab-tag.service.js
Comment (PyvesB) [reply services/gitlab/gitlab-tag.service.js:83]: I would tend to agree with the argument "let's be consistent with what the GitHub badges do".

Comment (calebcartwright) [reply services/gitlab/gitlab-tag.service.js:83]: Fair enough, I can go with that. One notable piece is that at past ops meetings we've discussed the need to brand our new GitLab badges with somewhat of a "beta" tag given the outstanding uncertainty around the auth story once we start to increase the scale of badges we offer and serve at runtime.

I'll post a note back in the main issue to that effect, which could in turn give us some leeway to revisit the message on this badge too.

**Comment by calebcartwright** [comment]
Nuts, should've just let ranger add a merge commit but did a rebase instead. One last call for a :+1: whenever anyone gets a chance :smile:

**Review by chris48s — APPROVED** [review]

**Pull request merged**
